Assets/Tests/InputSystem/Plugins/InputForUITests.cs

```diff
@@ -127,78 +110,389 @@ public void TouchIsPressedAndMouseEventsAreIgnored()
+    [TestCase(false)]
+    public void UI_Input_NavigationActionsWork(bool useProjectWideActionsAsset)
     {
-        MoveWithGamepad();
+        Update();
+        if (!useProjectWideActionsAsset)
+        {
+            // Remove the project-wide actions asset in play mode and player.
+            // It will call InputSystem.onActionChange and re-set InputSystemProvider.actionAsset
+            // This the case where no project-wide actions asset is available in the project.
+            InputSystem.s_Manager.actions = null;
+        }
+        Update();
 
-        Assert.IsTrue(m_InputForUIEvents.Count == 1);
-        Assert.That(m_InputForUIEvents[0] is Event
+        var gamepad = InputSystem.AddDevice<Gamepad>();
+        PressUpdateReleaseUpdate(gamepad.leftStick.left);
+        PressUpdateReleaseUpdate(gamepad.leftStick.up);
+        PressUpdateReleaseUpdate(gamepad.leftStick.right);
+        PressUpdateReleaseUpdate(gamepad.leftStick.down);
+
+        PressUpdateReleaseUpdate(gamepad.rightStick.left);
+        PressUpdateReleaseUpdate(gamepad.rightStick.up);
+        PressUpdateReleaseUpdate(gamepad.rightStick.right);
+        PressUpdateReleaseUpdate(gamepad.rightStick.down);
+
+        PressUpdateReleaseUpdate(gamepad.dpad.left);
+        PressUpdateReleaseUpdate(gamepad.dpad.up);
+        PressUpdateReleaseUpdate(gamepad.dpad.right);
+        PressUpdateReleaseUpdate(gamepad.dpad.down);
+
+        for (var i = 0; i < 3; ++i)
+        {
+            Assert.That(m_InputForUIEvents[i * 4] is
+            {
+                type: Event.Type.NavigationEvent,
+                asNavigationEvent: { type: NavigationEvent.Type.Move, direction: NavigationEvent.Direction.Left, eventSource: EventSource.Gamepad }
+            });
+
+            Assert.That(m_InputForUIEvents[i * 4 + 1] is
+            {
+                type: Event.Type.NavigationEvent,
+                asNavigationEvent: { type: NavigationEvent.Type.Move, direction: NavigationEvent.Direction.Up, eventSource: EventSource.Gamepad }
+            });
+
+            Assert.That(m_InputForUIEvents[i * 4 + 2] is
+            {
+                type: Event.Type.NavigationEvent,
+                asNavigationEvent: { type: NavigationEvent.Type.Move, direction: NavigationEvent.Direction.Right, eventSource: EventSource.Gamepad }
+            });
+
+            Assert.That(m_InputForUIEvents[i * 4 + 3] is
+            {
+                type: Event.Type.NavigationEvent,
+                asNavigationEvent: { type: NavigationEvent.Type.Move, direction: NavigationEvent.Direction.Down, eventSource: EventSource.Gamepad }
+            });
+        }
+
+
+        var joystick = InputSystem.AddDevice<Joystick>();
+        PressUpdateReleaseUpdate(joystick.stick.left);
+        PressUpdateReleaseUpdate(joystick.stick.up);
+        PressUpdateReleaseUpdate(joystick.stick.right);
+        PressUpdateReleaseUpdate(joystick.stick.down);
+
+        Assert.That(m_InputForUIEvents[12] is
```

Comment:
All of these specific offsets into event buffer, how likely do you think they are to change and break this test? I notice we do not check in-between events, would it make more sense to assert in-between each actuation above?

Reply:
Good points, I'll look into this